reports/glides_validator.livemd

```diff
@@ -145,10 +201,426 @@ table_ui =
+#   # Next 2 predictions:
+#   # First, group trip updates by timestamp.
+#   trip_updates
+#   |> Enum.group_by()
+#   # Apply to a stop. So group predictions by stop
+#   # Each stop gets 2 predictions
+#
+# end
+#
+# clamp_to_next_2_predictions =
+#   if limit_to_next_2_predictions do
+#     &do_clamp.(&1, stop_id)
+#   else
+#     &Function.identity/1
+#   end
+```
+
+## Per-Hour Counts of trips for which RTR made departure predictions vs. actual departures
 
 ```elixir
-Kino.ETS.new(:VehiclePositions)
+unix_timestamp_to_local_hour = fn timestamp ->
+  timestamp
+  |> DateTime.from_unix!()
+  |> DateTime.shift_zone!("America/New_York")
+  |> then(& &1.hour)
+end
+
+# QUESTION:
+# Do we need to worry about cases where actual departure happened around the start/end of the hour?
+# Maybe give a little wiggle room into the next/previous hour, for predictions..?
+
+predicted_departures_by_hour =
+  trip_updates
+  # Group by hour of the predicted departure (not hour the prediction was generated!),
+  # in local time.
+  # This requires splitting each trip update into its individual stop_time_update items.
+  |> Stream.flat_map(fn tr_upd ->
+    Enum.map(
+      tr_upd.trip_update.stop_time_update,
+      &%{trip_id: tr_upd.id, departure_time: &1.departure.time}
+    )
+  end)
+  |> Enum.group_by(
+    &unix_timestamp_to_local_hour.(&1.departure_time),
+    & &1.trip_id
+  )
+  |> Map.new(fn {hour, trip_ids} -> {hour, MapSet.new(trip_ids)} end)
+
+# |> IO.inspect(label: "predicted")
+
+actual_departures_by_hour =
+  vehicle_positions
+  # Group by hour
+  |> Enum.group_by(
+    &unix_timestamp_to_local_hour.(&1.vehicle.timestamp),
+    & &1.vehicle.trip.trip_id
+  )
+  |> Map.new(fn {hour, trip_ids} -> {hour, MapSet.new(trip_ids)} end)
+
+# |> IO.inspect(label: "actual")
+
+0..23
+|> Enum.map(fn hour ->
+  predicted_departure_trips = Map.get(predicted_departures_by_hour, hour, MapSet.new())
+  actual_departure_trips = Map.get(actual_departures_by_hour, hour, MapSet.new())
+
+  %{
+    "hour" => "#{pad.(hour)}:00",
+    "# of predicted departure trips" => MapSet.size(predicted_departure_trips),
+    "# of actual departure trips" => MapSet.size(actual_departure_trips)
+  }
+end)
+|> Kino.DataTable.new(
+  keys: ["hour", "# of predicted departure trips", "# of actual departure trips"]
+)
+
+# TODO: Rotate results so they start at beginning of the service day, rather than midnight?
+```
+
+## Per-Hour % of trips for which RTR made departure predictions vs. actual departures
+
+```elixir
+0..23
+|> Enum.map(fn hour ->
+  predicted_departures = Map.get(predicted_departures_by_hour, hour, MapSet.new())
+  actual_departures = Map.get(actual_departures_by_hour, hour, MapSet.new())
+
+  # *** For reviewer: ***
+  # This does trip matching, which the ticket description said is not needed.
+  # I'm not sure how else to interpret the requested metric, though--am I
+  # misunderstanding it, or is the ticket description incorrect?
```

Comment:
I think what's being referred to by trip matching is matching the trip on the prediction vs the trip on the actual departure. So if we predict trips A and B will depart stop X, and vehicles on trips C and D actually, we could that as 2 predictions and 2 departures, making it appear that we correctly anticipated all departures, as opposed to having 2 unanticipated departures, and 2 departures we predicted but didn't happen.

The idea being that we can't rely on the Glides trip matching the OCS trip, but the rider doesn't care what we call the trip, as long as a train is there.

Reply:
> matching the trip on the prediction vs the trip on the actual departure

Right—my code currently does that.
In this expr:

```ex
actual_departures_with_predictions =
  MapSet.intersection(predicted_departures, actual_departures)
```

The two sets are sets of trip IDs, which I match up by intersection, so I still think I'm doing it wrong.

Should I maybe instead be building a set of all the predicted departure _times_, and comparing that with a set of the actual departure _times_?
I think that lines up with the plain-English description of the ask.

This runs into the problem I raised the other day of what to do about predicted departure times changing over time, but I should be able to treat each change as a separate prediction as a few people in that thread proposed.

Reply:
I'd suggest conferring with @fsaid90 on Monday (and maybe demoing the reports as-is for him), but I think the idea with the percent, is that it's a column on the first table, rather than it's own that's simply %=predicted/actual. So if we predicted 11 departures and there were 10, we predicted 110% of the actual departures, 10% of the departures we predicted never occurred, on the other hand if we only predicted 9, we predicted 90%, and 10% were unexpected.